pep-0544.txt

```diff
@@ -0,0 +1,968 @@
+========================
+
+The ideas in this section were previously discussed in [several]_
+[discussions]_ [elsewhere]_.
+
+Make every class a protocol by default
+--------------------------------------
+
+Some languages such as Go make structural subtyping the only or the primary
+form of subtyping. We could achieve a similar result by making all classes
+protocols by default (or even always). However we believe that it is better
+to require classes to be explicitly marked as protocols, for the following
+reasons:
+
+* Protocols don't have some properties of regular classes. In particular,
+  ``isinstance()``, as defined for normal classes, is based on the nominal
+  hierarchy. In order to make everything a protocol by default, and have
+  ``isinstance()`` work would require changing its semantics,
+  which won't happen.
+* Protocol classes should generally not have many method implementations,
+  as they describe an interface, not an implementation.
+  Most classes have many implementations, making them bad protocol classes.
+* Experience suggests that many classes are not practical as protocols anyway,
+  mainly because their interfaces are too large, complex or
+  implementation-oriented (for example, they may include de facto
+  private attributes and methods without a ``__`` prefix).
+* Most actually useful protocols in existing Python code seem to be implicit.
+  The ABCs in ``typing`` and ``collections.abc`` are rather an exception, but
+  even they are recent additions to Python and most programmers
+  do not use them yet.
+* Many built-in functions only accept concrete instances of ``int``
+  (and subclass instances), and similarly for other built-in classes. Making
+  ``int`` a structural type wouldn't be safe without major changes to the
+  Python runtime, which won't happen.
+
+
+Support optional protocol members
+---------------------------------
+
+We can come up with examples where it would be handy to be able to say
+that a method or data attribute does not need to be present in a class
+implementing a protocol, but if it is present, it must conform to a specific
+signature or type. One could use a ``hasattr()`` check to determine whether
+they can use the attribute on a particular instance.
+
+Languages such as TypeScript have similar features and
+apparently they are pretty commonly used. The current realistic potential
+use cases for protocols in Python don't require these. In the interest
+of simplicity, we propose to not support optional methods or attributes.
+We can always revisit this later if there is an actual need.
+
+
+Make protocols interoperable with other approaches
+--------------------------------------------------
+
+The protocols as described here are basically a minimal extension to
+the existing concept of ABCs. We argue that this is the way they should
+be understood, instead of as something that replaces Zope interfaces,
+for example. Attempting such interoperabilities will significantly
```

Comment:
Zope interfaces are a *superset* of Protocols defined here, just using an incompatible syntax to define them. They necessarily use Attributes because pre-PEP 526 there was no other way to annotate this. Arguably in the 3.6+ world `zope.interface` could (optionally?) adopt the Protocol syntax.

In which case to achieve interoperability we'd only need to figure out a way to inform the type checker that "Interface" (or rather "InterfaceClass") is essentially "Protocol". This way we could type check successfully (pretty much ignoring `implements()` and treating those classes as *implicit* implementations. I think this is perfectly fine as the goal is **gradual typing**: if the type checker stays silent in some cases, that's OK. It just never should raise false positives.

Adapters, invariants, and runtime implements/provides checkers are obviously left out. This is a non-goal.